jsonschema-core.xml

```diff
@@ -800,49 +800,257 @@
             </t>
             <t>
-                If this URI identifies a retrievable resource, that resource SHOULD be of
-                media type "application/schema+json".
+                Meta-schema authoring is an advanced usage of JSON Schema, so the design of
+                meta-schema features emphasizes flexibility over simplicity.
             </t>
             <t>
-                The "$schema" keyword SHOULD be used in a root schema.
-                It MUST NOT appear in subschemas.
+                The role of a vocabulary is to declare which keywords (including sub-keywords
+                such as those in JSON Hyper-Schema's Link Description Object) are in use,
+                and with which semantics.  The semantics are indicated by the document
+                that publicizes the vocabulary URI.  At this time, there is no machine-readable
+                description of keywords other than validation rules, which appear in the
+                meta-schema.
             </t>
             <t>
-                <cref>
-                    Using multiple "$schema" keywords in the same document would imply that the
-                    vocabulary and therefore behavior can change within a document.  This would
-                    necessitate resolving a number of implementation concerns that have not yet
-                    been clearly defined.  So, while the pattern of using "$schema" only in root
-                    schemas is likely to remain the best practice for schema authoring,
-                    implementation behavior is subject to be revised or liberalized in
-                    future drafts.
-                </cref>
-                <!--
-                    In particular, the process of validating an instance, including validating a
-                    schema as an instance against its meta-schema, only allows for a single set
-                    of rules across the entire instance document.  There is no equivalent of
-                    changing the meta-schema partway through the validation for non-schema
-                    instances.
-                  -->
+                Meta-schemas are separate from vocabularies to allow for the same sets of
+                vocabularies to be combined in different ways, and for meta-schema authors
+                to impose additional constraints such as forbidding certain keywords, or
+                performing unusually strict syntactical validation, as might be done
+                during a development and testing cycle.
             </t>
             <t>
-                Values for this property are defined in other documents and by other parties.
-                JSON Schema implementations SHOULD implement support for current and previous
-                published drafts of JSON Schema vocabularies as deemed reasonable.
+                A schema is considered to be processed as meta-schema if it was
```

Comment:
How would processing of a meta-schema differ from processing any other schema?  I ask because Manatee.Json processes all of the *draft-0x* meta-schemas like any other schema (partially to ensure self-validation).

Reply:
At this point, it's mostly just "if it's a meta-schema, you MUST stop processing if you don't recognize a required vocabulary in `$vocabulary`, otherwise you MUST ignore `$vocabulary`"

If you have a meta-schema M that requires vocabs A, B, and C, and it's own meta-schema M' requires vocabs A, C, and D (meaning that M itself uses A, C, and D, even though it describes A, , and C), then you need to be able to validate M without understanding B.  M describes B, but does not require it.  M' requires D, even though M only uses but does not describe it.

Reply:
I say "mostly" because I remember AJV doing something else different, but I think that had to do with it getting confused about needing the meta-schema already validated before it could validate the meta-schema or some other weird circular thing that I don't think the spec requires.

Reply:
Another question: What would signify that a schema is a meta-schema so that an implementation would know to process it differently?

Reply:
As the line you attached this thread to says:

>  A schema is considered to be processed as meta-schema if it was identified as such in a schema's "$schema".  Implementations MAY also offer alternate ways to process schemas using meta-schema rules.

So if you are processing a schema (which is generally how things start), and you follow a `$schema` URI to another document, that document is a meta-schema in this context because it was the direct target of `$schema`.  So `$vocabulary` is respected in its root schema (but nowhere else).

Reply:
So the only way to know whether a schema is a meta-schema is to know if it's used as a meta-schema by another schema? (Say "schema" some more.)  What if I start with the meta-schema?  Is it just processed like any other schema?

Reply:
As the text says:

> Implementations MAY also offer alternate ways to process schemas using meta-schema rules.

I don't mean to be a jerk here, but I don't understand what's not clear.  There are two ways that I know of that implementations process meta-schemas:

* They read `"$schema"` and either fetch the resource or load it from cache.  You obviously know it is a meta-schema because... `"$schema"`  Is there anything confusing about this part?
* There is some way to clearly hand a meta-schema to an implementation and tell it to treat it that way.  I don't know why you would do that, how, or what purpose it would serve.  The point is to allow an out for people who want to take it because, I dunno, they process meta-schemas up-front and decide whether they are supported by checking `$vocabulary`, and then don't process them as individual schemas come in.  I don't know, and the point is that I'm not preventing anyone from doing whatever they need like that.

If you start with the meta-schema as your schema, then it would not seem to fall into either of those two categories, so yes, it's being treated as a schema.

Reply:
@gregsdennis awesome, I will try to figure out how to clarify that point.

Thanks for all of your feedback, I knew posting it that it would be too confusing but I couldn't figure out how to make it better on my own.

Reply:
@gregsdennis I think I have fixed/clarified this, please let me know / mark the conversation resolved if the new text works.

Reply:
I can't mark it resolved.